### PR TITLE
Migrate from Knex to Kysely

### PR DESCRIPTION
**Pull request by Brainicism**
https://github.com/kysely-org/kysely

### CODE REVIEW

**Review by taahamahdi — CHANGES_REQUESTED** [review]

File: src/commands/game_commands/leaderboard.ts [hunk 1]
```diff
@@ -671,136 +706,152 @@ export default class LeaderboardCommand implements BaseCommand {
...
-                                                    player.exp
-                                                )} EXP`;
-
-                                                value = `${expGained} | ${level}`;
-                                            }
-
-                                            break;
-                                        case LeaderboardType.GAMES_PLAYED: {
-                                            const games = i18n.translate(
+                                topPlayers.flatMap(
```
[the remark this answers is not in the document]
Comment (Brainicism) [reply src/commands/game_commands/leaderboard.ts:709]: no i cant

File: src/commands/game_commands/upcomingreleases.ts [hunk 2]
```diff
@@ -34,8 +30,8 @@ const FIELDS_PER_EMBED = 9;
 interface UpcomingRelease {
     name: string;
     artistName: string;
-    hangulArtistName?: string;
-    releaseType: ReleaseType;
+    hangulArtistName: string | null;
+    releaseType: "album" | "ep" | "single" | "undefined";
```
[the remark this answers is not in the document]
Comment (Brainicism) [reply src/commands/game_commands/upcomingreleases.ts:34]: No, not possible. I am planning to kill all enums in favor of union types in a future PR anyway
https://www.reddit.com/r/typescript/comments/103nino/whats_up_with_all_the_enum_hate_lately/

https://www.typescriptlang.org/docs/handbook/enums.html#objects-vs-enums

File: src/enums/option_types/gender.ts [hunk 3]
```diff
@@ -1,8 +1,9 @@
-enum Gender {
-    MALE = "male",
-    FEMALE = "female",
-    COED = "coed",
-    ALTERNATING = "alternating",
-}
+export const availableGenders = [
```
[the remark this answers is not in the document]
Comment (Brainicism) [reply src/enums/option_types/gender.ts:1]: See above

File: src/helpers/game_utils.ts [hunk 4]
```diff
@@ -173,33 +182,37 @@ export async function getMatchingGroupNames(
...
-    ).map((x) => ({ id: x.id, name: x.name }));
+    const artistIds = (
+        await dbContext.kpopVideos
+            .selectFrom("app_kpop_group")
+            .select(["id"])
+            .where("name", "in", rawGroupNames)
+            .execute()
+    ).map((x) => x.id);
+
+    const matchingGroups = artistIds.length
```
[the remark this answers is not in the document]
Comment (Brainicism) [reply src/helpers/game_utils.ts:193]: i think its fine

File: src/helpers/game_utils.ts [hunk 5]
```diff
@@ -471,40 +500,46 @@ export async function updatePremium(
...
-        .onConflict(["user_id", "badge_name"])
-        .ignore();
+    const activePatronsPayload = activePatrons.map((x) => ({
+        active: x.activePatron ? 1 : 0,
+        first_subscribed: x.firstSubscribed,
+        user_id: x.discordID,
+        source: "patreon" as const,
+    }));
+
+    await Promise.all(
```
[the remark this answers is not in the document]
Comment (Brainicism) [reply src/helpers/game_utils.ts:510]: not sure how to do this and do onDuplicateKeyUpdate at the same time, will revisit this later

File: src/helpers/management_utils.ts [hunk 6]
```diff
@@ -285,16 +300,17 @@ export async function reloadBonusGroups(): Promise<void> {
...
+    const artistAliasMapping = await dbContext.kmq
+        .selectFrom("available_songs")
         .select([
             "artist_name_en",
             "artist_name_ko",
             "artist_aliases",
             "id_artist",
         ])
-        .whereRaw("artist_name_en NOT LIKE ?", ["%+%"]);
+        .distinct()
```
[the remark this answers is not in the document]
Comment (Brainicism) [reply src/helpers/management_utils.ts:311]: mysql only supports distinct on all selected fields. this new query is functionally equivalent to the previous

File: src/scripts/download-new-songs.ts [hunk 7]
```diff
@@ -213,63 +215,75 @@ const downloadSong = (db: DatabaseContext, id: string): Promise<void> => {
...
+                    "app_kpop_group",
+                    "app_kpop.id_artist",
+                    "app_kpop_group.id"
+                )
                 .select([
-                    "app_kpop.name AS songName",
-                    "app_kpop_group.name AS artistName",
-                    "vlink AS youtubeLink",
-                    "app_kpop.views AS views",
-                    "app_kpop.tags AS tags",
```
[the remark this answers is not in the document]
Comment (Brainicism) [reply src/scripts/download-new-songs.ts:226]: We don't need it. Removed tags from the av one

File: src/scripts/store-daily-stats.ts [hunk 8]
```diff
@@ -7,44 +7,45 @@ const storeDailyStats = async (serverCount: number): Promise<void> => {
...
-        await dbContext
-            .kmq("player_stats")
-            .where("first_play", ">=", dateThreshold)
-            .count("* as count")
-    )[0].count;
+    const recentGameSessions = (await dbContext.kmq
+        .selectFrom("game_sessions")
+        .where("start_date", ">", dateThreshold)
+        .select((eb) => eb.fn.countAll<number>().as("count"))
+        .executeTakeFirst())!.count;
```
[the remark this answers is not in the document]
Comment (Brainicism) [reply src/scripts/store-daily-stats.ts:14]: Realistically wouldnt happen on prod but added them anyway

File: src/seed/bootstrap.ts [hunk 9]
```diff
@@ -73,47 +77,117 @@ async function songThresholdReached(db: DatabaseContext): Promise<boolean> {
...
+ */
+export async function performMigrations(db: DatabaseContext): Promise<void> {
+    logger.info("Performing migrations (up)...");
+    const migrator = new Migrator({
+        db: db.kmq,
+        provider: new FileMigrationProvider({
+            fs: fsp,
+            path,
+            // This needs to be an absolute path.
+            migrationFolder: path.join(__dirname, "../migrations"),
```
[the remark this answers is not in the document]
Comment (Brainicism) [reply src/seed/bootstrap.ts:114]: yes

File: src/seed/bootstrap.ts [hunk 10]
```diff
@@ -137,9 +211,10 @@ async function bootstrapDatabases(): Promise<void> {
     }
 
     logger.info("Cleaning up stale data");
-    await db.kmq.raw(
-        "SELECT * FROM system_stats WHERE date < DATE(NOW() - INTERVAL 3 MONTH)"
```
[the remark this answers is not in the document]
Comment (Brainicism) [reply src/seed/bootstrap.ts:142]: yes

File: src/seed/seed_db.ts [hunk 11]
```diff
@@ -375,7 +409,9 @@ async function validateSqlDump(
                 `mysql -u ${process.env.DB_USER} -p${process.env.DB_PASS} -h ${process.env.DB_HOST} --port ${process.env.DB_PORT} kpop_videos_validation < ${validationDedupGroupNamesSqlPath}`
             );
 
-            await db.kpopVideosValidation.raw("CALL DeduplicateGroupNames();");
+            await sql
+                .raw("CALL DeduplicateGroupNames();")
```
[the remark this answers is not in the document]
Comment (Brainicism) [reply src/seed/seed_db.ts:413]: sql`[x]` can be used part of builder expressions on existing tables. sql.raw for arbitrary queries

File: src/seed/seed_db.ts [hunk 12]
```diff
@@ -644,6 +693,10 @@ async function seedAndDownloadNewSongs(db: DatabaseContext): Promise<void> {
         try {
             await loadStoredProcedures();
             await seedAndDownloadNewSongs(db);
+
+            if (process.env.NODE_ENV !== EnvType.PROD) {
```
[the remark this answers is not in the document]
Comment (Brainicism) [reply src/seed/seed_db.ts:694]: prod doesnt need the typings

File: src/structures/guild_preference.ts [hunk 13]
```diff
@@ -322,11 +330,12 @@ export default class GuildPreference {
...
     async listPresets(): Promise<string[]> {
         const presets = (
-            await dbContext
-                .kmq("game_option_presets")
+            await dbContext.kmq
+                .selectFrom("game_option_presets")
                 .select(["preset_name"])
                 .where("guild_id", "=", this.guildID)
-                .distinct("preset_name")
+                .distinct()
```
[the remark this answers is not in the document]
Comment (Brainicism) [reply src/structures/guild_preference.ts:337]: see above

File: src/fact_generator.ts [hunk 14]
```diff
@@ -226,12 +229,7 @@ async function recentMusicShowWin(lng: LocaleType): Promise<string[]> {
         ])
         .where("date", ">", twoWeeksPriorDate)
         .where("app_kpop_ms.id_musicvideo", "!=", 0)
-        .join("app_kpop_group", function join() {
-            this.on("app_kpop_ms.id_artist", "=", "app_kpop_group.id");
-        })
-        .join("app_kpop", function join() {
-            this.on("app_kpop_ms.id_musicvideo", "=", "app_kpop.vlink");
```
[the remark this answers is not in the document]
Comment (Brainicism) [reply src/fact_generator.ts:233]: yes

File: src/fact_generator.ts [hunk 15]
```diff
@@ -620,26 +622,27 @@ async function mostViewedSoloArtist(lng: LocaleType): Promise<string[]> {
...
+    const result = await dbContext.kpopVideos
+        .selectFrom("app_kpop")
+        .innerJoin("app_kpop_group", "app_kpop.id_artist", "app_kpop_group.id")
         .select(["app_kpop_group.issolo as issolo"])
-        .join("app_kpop_group", function join() {
-            this.on("app_kpop.id_artist", "=", "app_kpop_group.id");
-        })
-        .groupBy("app_kpop_group.issolo")
-        .sum("app_kpop.views as views")
-        .orderBy("issolo", "DESC")
```
[the remark this answers is not in the document]
Comment (Brainicism) [reply src/fact_generator.ts:631]: yes

**Review by taahamahdi — APPROVED** [review]

**Pull request merged**